### PR TITLE
ec2.securitygroup: Avoid late init if it seems unsafe

### PR DESCRIPTION
**Pull request by chlunde**
If the structure of the ingress/egress rules are different, or any port
numbers/protocols do not match, avoid late init because we will not be
able to do it safely. The array does not have a unique key.

Also avoid late init of description because I have not seen that AWS
sets description.

Updates #618

### Description of your changes

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

`kubectl create` this object with description for one IP range:
```yaml
apiVersion: ec2.aws.crossplane.io/v1beta1
kind: SecurityGroup
metadata:
  name: sg-1
spec:
  forProvider:
    description: desc
    groupName: sg1
    region: eu-north-1
    ingress:
    - ipProtocol: "tcp"
      fromPort: 2
      toPort: 2
      ipRanges:
      - cidrIp: 10.0.0.0/8
    - ipProtocol: "tcp"
      fromPort: 1
      toPort: 1
      ipRanges:
      - cidrIp: 10.0.0.0/8
        description: "foo"
    vpcId: vpc-...
```

next, `kubectl apply` with a completely different rule, and without description:

```yaml
apiVersion: ec2.aws.crossplane.io/v1beta1
kind: SecurityGroup
metadata:
  name: sg-1
spec:
  forProvider:
    description: desc
    groupName: sg1
    region: eu-north-1
    ingress:
    - ipProtocol: "tcp"
      fromPort: 2
      toPort: 2
      ipRanges:
      - cidrIp: 10.0.0.0/8
    - ipProtocol: "tcp"
      fromPort: 3
      toPort: 3
      ipRanges:
      - cidrIp: 172.0.0.0/8
    vpcId: vpc-...
```

With the old code, using `kubectl get -o yaml` we see now that the `description: "foo"` from 10.0.0.0/8:1 is now set in `172.0.0.0/8:3`. This is no longer the case.

```yaml
    - fromPort: 3
      ipProtocol: tcp
      ipRanges:
      - cidrIp: 172.0.0.0/8
        description: foo
      toPort: 3
```

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: pkg/clients/ec2/securitygroup.go [hunk 1]
```diff
@@ -198,42 +199,55 @@ func LateInitializeIPPermissions(spec []v1beta1.IPPermission, o []ec2.IpPermissi
...
+	for i := range o {
+		spec[i].FromPort = awsclients.LateInitializeInt64Ptr(spec[i].FromPort, o[i].FromPort)
+		spec[i].ToPort = awsclients.LateInitializeInt64Ptr(spec[i].ToPort, o[i].ToPort)
+		spec[i].IPProtocol = awsclients.LateInitializeString(spec[i].IPProtocol, o[i].IpProtocol)
+
 		for j := range o[i].UserIdGroupPairs {
-			if len(spec[i].UserIDGroupPairs) == j {
-				spec[i].UserIDGroupPairs = append(spec[i].UserIDGroupPairs, v1beta1.UserIDGroupPair{})
-			}
 			spec[i].UserIDGroupPairs[j].Description = awsclients.LateInitializeStringPtr(
```
Comment (chlunde) [inline pkg/clients/ec2/securitygroup.go:251]: Is there any more checks we should do here for UserIDGroupPairs

File: pkg/clients/ec2/securitygroup.go [hunk 2]
```diff
@@ -2,6 +2,7 @@ package ec2
 
 import (
 	"encoding/json"
+	"log"
```
Comment (chlunde) [inline pkg/clients/ec2/securitygroup.go:5]: if you like this we can remove the logging and move it to a new helper method

**Comment by chlunde** [comment]
what do you think @muvaf

**Comment by chlunde** [comment]
I can write a unit test if you agree that this change is needed.

**Comment by muvaf** [comment]
From what I can tell, we rely on `protocol`, `fromPort` and `toPort` as key in this [other PR](https://github.com/crossplane/provider-aws/pull/631). Would we be able to rely to those as keys here, too?

**Pull request closed without merging**